torch/utils/data/dataloader.py

```diff
@@ -485,14 +523,74 @@ def __iter__(self) -> _BaseDataLoaderIter:
         # However, in the case of a multiple workers iterator
         # the iterator is only created once in the lifetime of the
         # DataLoader object so that workers can be reused
-        if self.persistent_workers and self.num_workers > 0:
+        loading_state = self.stateful and self.next_iter_state is not None
+        if self._initial_iter_for_state_dict:
+            self._initial_iter_for_state_dict = False
+            assert self._iterator is not None
+        elif self.persistent_workers and self.num_workers > 0:
             if self._iterator is None:
                 self._iterator = self._get_iterator()
             else:
                 self._iterator._reset(self)
-            return self._iterator
+        elif (
+            loading_state
+            and self.next_iter_state
+            and self.next_iter_state.get(_StateKeys.ITERATOR_FINISHED, False)
+        ):
+            self.next_iter_state = None
+            self._iterator = self._get_stateful_iterator()
```

Comment:
Add comment because I'm still confused by this
```suggestion
            # If the previously saved state was in the finished state, we want to load the start of the next epoch
            # instead of the end of the previous. See https://github.com/meta-pytorch/data/issues/1437
            self.next_iter_state = None
            self._iterator = self._get_stateful_iterator()
```